Allow SecurityComponent to check fields that are optional, but must have a specific value if included (issue 7848)

Zuluru opened the pull request:
See [issue 7848](https://github.com/cakephp/cakephp/issues/7848) for discussion.

I haven't yet made any changes in the documentation, pending acceptance of general implementation of this PR. If these changes look good but you want the docs updated before merging, let me know and I can take care of that.

Commit:
Sqlschema Fix for views and Bake

Zuluru committed:
Fix docblocks

Zuluru committed:
Fix docblocks

Zuluru committed:
Use recommended parameter order

Zuluru committed:
Use recommended parameter order

Zuluru committed:
Add 'optional' option for secure parameter to form inputs

Zuluru committed:
Add 'optional' option for secure parameter to form inputs

thinkingmedia commented on src/Controller/Component/SecurityComponent.php:
```diff
@@ -496,7 +560,7 @@ protected function _debugPostTokenNotMatching(Controller $controller, $hashParts
     {
         $messages = [];
```
don't think this was even needed in the original.

Zuluru replied:
Yes, it is. Line 569 appends to the array.

thinkingmedia replied:
ah it was hidden by the merge summary.

Zuluru committed:
Change else if to elseif

Zuluru committed:
Change else if to elseif

Zuluru committed:
Added missing parameter to docblock

Zuluru committed:
Added missing parameter to docblock

markstory commented on src/Controller/Component/SecurityComponent.php:
```diff
@@ -486,6 +490,66 @@ protected function _sortedUnlocked($data)
+            if (!empty($optional)) {
+                $optional = explode('|', $optional);
+                foreach ($optional as $field) {
+                    list($name, $hash) = explode('=', $field);
```
This looks like we're rebuilding key/value encoding. Any reason to not use json encoding?

Zuluru replied:
Yeah, that should simplify things. I was sticking with the | delimiting that the other params used, but there's no compelling reason to do so.

markstory commented on src/Controller/Component/SecurityComponent.php:
```diff
@@ -486,6 +490,66 @@ protected function _sortedUnlocked($data)
+    {
+        $fields = $controller->request->data;
+        unset($fields['_Token'], $fields['_csrfToken']);
+        $fields = Hash::flatten($fields);
```
Any way we could reduce the duplication here and when the required fields are validated?

Zuluru replied:
Do you mean the three common lines between `_validateOptionalFields` and `_debugPostTokenNotMatching`, or the similarity in purpose between `_validateOptionalFields` and `_matchExistingFields`?

markstory commented on src/Controller/Component/SecurityComponent.php:
```diff
@@ -486,6 +490,66 @@ protected function _sortedUnlocked($data)
+        $optional = $this->_optional($controller->request->data);
+        foreach ($optional as $name => $hash) {
+            if (array_key_exists($name, $fields)) {
+                if ($hash !== Security::hash($name . $fields[$name] . Security::salt() . $token)) {
```
You need to specify the hash method here. People can change the default to an incompatible mode.

markstory replied:
Is there a reason you're manually inserting the salt vs. relying on the hash() method to append the salt?

Zuluru replied:
No **good** reason, no. Just followed the other `Security::hash` calls in here.

markstory commented on src/Controller/Component/SecurityComponent.php:
```diff
@@ -486,6 +490,66 @@ protected function _sortedUnlocked($data)
+            return $optionalFields;
+        } else {
+            return [];
+        }
```
I would invert the condition and return early here so you can avoid an else, and reduce the level of indentation here.

thinkingmedia replied:
```
   $optional = urldecode($data['_Token']['optional'] ?: '');
   if (empty($optional)) {
          return [];
   }
   $optionalFields = [];
   foreach (explode('|', $optional) as $field) {
        list($name, $hash) = explode('=', $field);
        $optionalFields[$name] = $hash;
   }
   return $optionalFields;
```

Just because....

Zuluru replied:
With the change to JSON encoding, it's even easier than that...

Zuluru committed:
Complete suggested changes to hashing and encoding

Zuluru committed:
Complete suggested changes to hashing and encoding

Zuluru commented:
I think I've addressed all the comments except one where I had a follow-up question. If you have a chance to respond some time this weekend, I should be able to finish up early next week.

Zuluru commented:
Is there anything else that I need to do on this? Any problem with the PR?

markstory commented:
@Zuluru I'll take another look this weekend.

Zuluru commented:
I've merged in 3.2.8, and the Travis CI is failing, but the failure seems to be nowhere near my changes?

thinkingmedia commented:
@Zuluru I know, I'm working on trying to fix it #8720

Zuluru commented:
Okay, thought I might have broken something. :-)

Zuluru commented:
@thinkingmedia I see the builds are working again. Any way to get it to re-run it on my PR, other than pushing another commit from my end? That red X is mocking me...

thinkingmedia commented:
@Zuluru the tests can only be re-run against the commit, which hasn't changed so you wouldn't get any different of a result. You need to update the PR to run the tests against the head of master.

Zuluru commented:
Had a chance to merge with the latest updates, and the tests pass again. Yay! @markstory, any updates on your end?

Zuluru commented:
Just noticed there was a conflict with 3.2.10, so it's been merged and resolved.

markstory committed:
Add config file for stickler-ci.com

Hopefully by making PHPCS errors more visible, we can spend less time
looking up why travis builds failed.

markstory committed:
Merge branch 'tgr0ss-master' into master

Enable SQLServer to reflect schema for Views in addition to tables.

Refs #7846
Refs #6501

ADmad committed:
Merge pull request #8954 from cakephp/add-stickler

Add config file for stickler-ci.com

Commit:
Mention `finder` option in the `Query::contain` docs.

markstory committed:
Merge pull request #8975 from ndm2/add-finder-option-to-contain-docs

Mention `finder` option in the `Query::contain` docs.

markstory committed:
Revert "Default output timezone v3"

lorenzo committed:
Merge pull request #8976 from cakephp/revert-8935-default-output-timezone-new

Revert "Default output timezone v3"

markstory committed:
Correct doc type.

Refs #8976

ADmad committed:
Try using xdebug to get coverage report.

ADmad committed:
No need to install apcu_bc on travis.

ADmad committed:
Merge pull request #8977 from cakephp/code-coverage

Try using xdebug to get coverage report.

chinpei215 committed:
Add support for per-column collation

steinkel committed:
fix docblock typo

dereuromark committed:
Merge pull request #8979 from cakephp/steinkel-patch-1

fix docblock typo

Commit:
Improving the error message for associations with no PK.

This makes it more clear what needs to be done to fix the issue.

markstory committed:
Merge pull request #8980 from burzum/better-error-msg

Improving the error message for associations with no PK.

dakota committed:
Throw a RuntimeException if the validation method does not return a validator

dakota committed:
Add @throws tag to docblock

dakota committed:
Appease the Stickler

dakota committed:
Add in some brackets

dakota committed:
Because windows is silly

lorenzo committed:
Merge pull request #8983 from dakota/make-validation-unsurprising

Throw a RuntimeException if the validation method does not return a validator

jeremyharris committed:
Added assertEventFired and assertEventFiredWith to TestCase

ravage84 committed:
Remove space before object operator

dereuromark committed:
Merge pull request #8987 from ravage84/space-before-object-operator

Remove space before object operator

ravage84 committed:
Don't use unset $translator, use given $formatter instead

ravage84 committed:
Add space before operators

lorenzo committed:
Merge pull request #8988 from ravage84/diffFormatter

Don't use unset $translator, use given $formatter instead

dereuromark committed:
Merge pull request #8989 from ravage84/space-before-object-operator

Add space before operators

ravage84 committed:
Let Stickler CI ignore the 2.x related branches

jeremyharris committed:
Throwing assertion error if no event list is configured

markstory committed:
Merge pull request #8991 from ravage84/patch-5

Let Stickler CI ignore the 2.x related branches

hyaza committed:
problem. In the session of Database use , it expires is long . (#8986)

Session using the database, will expire period is doubled in the gc.

Rather than the update time when you want to save the session , as specified by the expiration time .
Nevertheless , 'gc' at the time of , are we calculate the expire time .
A result , not expire at the correct time .
Indeed , the life of the fold session data time assumes are treated longer.

PHP default session management , use the date and time of the recording as it is .
However, in the moment of CakePHP adds the date and time of the expiration date to use as 'expire'.
Here 's a different point , but seems to change in the processing of as expire in the method of 'gc' is insufficient .

lorenzo committed:
Merge pull request #8985 from jeremyharris/feature-assert-events

[RFC] Add assertEventFired and assertEventFiredWith

thinkingmedia committed:
wrap findOrCreate with transactional

thinkingmedia committed:
Update Table.php

thinkingmedia committed:
don't update private variable unless repo exists. (#9009)

markstory committed:
Don't use empty()

Refs #9009

burzum committed:
Fix for #8942 customized redirect route class (#9010)

markstory committed:
Update version number to 3.2.11

markstory committed:
Rework cid: images (#8966)

* Adding CID support for HtmlHelper::image() #8877

* Move cid: checking in to UrlHelper.

While this does allow `cid:` prefixes to be used on all asset types, it
also enables other `a-z:` prefixes to be used easily as well.

Refs #8898

* Allow uppercase CID as well.

markstory committed:
Don't invoke rules twice.

By decorating rules we can detect and prevent duplicate invocation of
rules. This issue was most easily observed with isUnique and existsIn
rules with custom errorField options. Previously those rules would emit
2 errors instead of one.

Commit:
bug: target and junction conditions not reset properly

markstory committed:
Merge pull request #9012 from ulistaerk/master

bug: target and junction conditions not reset properly

markstory committed:
Fix PHPCS error.

markstory committed:
Forward the current connection to the generated table.

This helps make belongsToMany associations in plugins easier to work
with, as the generated table will inherit the source table's connection,
which is _generally_ what people will expect.

Refs #9016

lorenzo committed:
Merge pull request #9018 from cakephp/issue-9016

Forward the current connection to the generated table.

lorenzo committed:
Merge pull request #9017 from cakephp/issue-8993

Don't invoke rules twice.

markstory committed:
Only propagate the connection on automodels.

Do some refactoring in BelongsToMany::junction() to reduce nesting
depth.

Refs #9018

lorenzo committed:
Merge pull request #9021 from cakephp/issue-9018-follow-up

Only propagate the connection on automodels.

dakota committed:
Association formatter should not invoke the entity get method

lorenzo committed:
Merge pull request #9022 from dakota/no-extract-associations

Association formatter should not invoke the entity get method

markstory committed:
Don't test sslv3

Generally people are ceasing to support sslv3 as it has a number of
weaknesses. Lets not test against old busted ciphers.

chinpei215 committed:
Fix empty usem. Fix test cases.

ADmad committed:
Fix auto linking for URLs with uppercased protocol.

Closes #9033

dakota committed:
Fix issue with cache groups when creating a cache engine

If a cache engine is configured by passing in a already constructed instance, any potential groups used by that engine are lost. This can have dire consequences when trying to call (for example), `Cache::groupConfigs()`. I noticed this problem with DebugKit's cache spies.

markstory committed:
Merge pull request #8978 from chinpei215/per-column-collation

Add support for per-column collation

markstory committed:
Merge pull request #9034 from cakephp/issue-9033

Fix auto linking for URLs with uppercased protocol.

dakota committed:
Add test for cache instances and group config

dakota committed:
Fixes the failing test due to cache config change.

The extra config() call pushes the expected method out by one in the greater sequence of things.

lorenzo committed:
Merge pull request #9035 from dakota/dakota-patch-1

Fix issue with cache groups when creating a cache engine

Commit:
I forgot to return the result in findOrCreate, thank goodness for unit tests

lorenzo committed:
Merge pull request #9007 from thinkingmedia/patch-5

wrap findOrCreate with transactional

Zuluru committed:
Merge branch 'master' of github.com:Zuluru/cakephp

thinkingmedia commented:
why are there commits from master appearing on the change list of this PR?

Zuluru commented:
I probably merged it with 3.2.11 incorrectly. :-( Thought I followed the same steps as previously...

thinkingmedia commented:
I think it's because your PR is from a master. So what might be happening is that GitHub has synced the two masters which results in all the commits being shown in the PR.

Next time make a PR from a branch. If you want (but you'll loose the comment history). You can close this PR, create a branch and open a new PR for that branch. In the PR message just link back to this PR using the hashtag shortcut.

Zuluru commented:
Created new PR (9888) to resolve git management issues.
